examples/FileIO/WriteBytesToFile.dfy

```diff
@@ -0,0 +1,79 @@
+// RUN: %baredafny run --no-verify --target:java "%s" --input "%S/../../src/FileIO/FileIO.java" -- "%t_java" "java.nio.file.FileSystemException:" >> "%t"
+// RUN: %baredafny run --no-verify --target:js "%s" --input "%S/../../src/FileIO/FileIO.js" -- "%t_js" "Error: ENOENT" >> "%t"
+// TODO: %baredafny run --no-verify --target:py "%s" --input "%S/../../src/FileIO/FileIO.py" -- "%t_py" "[Errno 2]" >> "%t"
+// RUN: %diff "%s.expect" "%t"
+
+//// Check that written files match expectations
+// RUN: %diff "data.txt" "%t_cs/output_plain"
+// RUN: %diff "data.txt" "%t_cs/foo/bar/output_nested"
+// RUN: %diff "data.txt" "%t_cs/foo/output_up"
+// RUN: %diff "data.txt" "%t_java/output_plain"
+// RUN: %diff "data.txt" "%t_java/foo/bar/output_nested"
+// RUN: %diff "data.txt" "%t_java/foo/output_up"
+// RUN: %diff "data.txt" "%t_js/output_plain"
+// RUN: %diff "data.txt" "%t_js/foo/bar/output_nested"
+// RUN: %diff "data.txt" "%t_js/foo/output_up"
+// TODO: %diff "data.txt" "%t_py/output_plain"
+// TODO: %diff "data.txt" "%t_py/foo/bar/output_nested"
+// TODO: %diff "data.txt" "%t_py/foo/output_up"
+
+include "../../src/FileIO/FileIO.dfy"
+
+module WriteBytesToFile {
+  import FileIO
+
+  method Main(args: seq<string>) {
+    expect |args| > 0;
+    expect |args| == 3, "usage: " + args[0] + " OUTPUT_DIR EXPECTED_ERROR_PREFIX";
+    var outputDir := args[1];
+    var expectedErrorPrefix := args[2];
+
+    // Happy paths: write files to the output dir. (The %diff LIT commands check that we wrote the correct content.)
+    {
+      // Ideally we would define `str` as a constant and compute `bytes` automatically.
+      // To do so, we would need to convert each `char` in `str` to a `bv8` value, by using `as bv8`.
+      // But that triggers a bug in the Java compiler: <https://github.com/dafny-lang/dafny/issues/2942>.
+      // So for now we settle for manually writing out the desired `bytes`,
+      // and statically verifying that these values match the characters of `str`.
+      ghost var str := "Hello world\nGoodbye\n";
+      var bytes: seq<bv8> := [
+        // "Hello world\n"
+        0x48, 0x65, 0x6c, 0x6c, 0x6f, 0x20, 0x77, 0x6f, 0x72, 0x6c, 0x64, 0x0a,
+        // "Goodbye\n"
+        0x47, 0x6f, 0x6f, 0x64, 0x62, 0x79, 0x65, 0x0a
+      ];
+      assert forall i | 0 <= i < |bytes| :: bytes[i] as int == str[i] as int;
+
+      // Write directly into the output directory
+      {
+        var res := FileIO.WriteBytesToFile(outputDir + "/output_plain", bytes);
+        expect res.Success?, "unexpected failure writing to output_plain: " + res.error;
+      }
+      // Ensure that nonexistent parent directories are created as necessary
+      {
+        var res := FileIO.WriteBytesToFile(outputDir + "/foo/bar/output_nested", bytes);
+        expect res.Success?, "unexpected failure writing to output_nested: " + res.error;
+      }
+      // Ensure that file paths with .. are handled correctly
```

Comment:
Worth mentioning that paths with `..` should always work in the top level API documentation?

Reply:
Good call, fixed in latest commits.